server/app/auth/oidc/applicant/OidcApplicantProfileAdapter.java

```diff
@@ -0,0 +1,168 @@
+ * accounts - that's not usually needed in web applications which is why we have to write this class
+ * - pac4j doesn't come with it. It's abstract because AD and IDCS need slightly different
+ * implementations of the two abstract methods.
+ */
+public abstract class OidcApplicantProfileAdapter extends OidcProfileAdapter {
+
+  protected final Config app_configuration;
+
+  protected final String localeAttributeConfigName = "locale_attribute";
+  protected final String firstNameAttributeConfigName = "first_name_attribute";
+  protected final String secondNameAttributeConfigName = "second_name_attribute";
+  protected final String emailAttributeConfigName = "email_attribute";
+
+  public OidcApplicantProfileAdapter(
+      OidcConfiguration configuration,
+      OidcClient client,
+      Config app_configuration,
+      ProfileFactory profileFactory,
+      Provider<UserRepository> applicantRepositoryProvider) {
+    super(configuration, client, profileFactory, applicantRepositoryProvider);
+    this.app_configuration = app_configuration;
+  }
+
+  /*
+   * Provide the prefix used in the application.conf
+   */
+  protected abstract String attributePrefix();
+
+  protected String getConfigurationValue(String attr) {
+    return getConfigurationValue(attr, "");
+  }
+
+  protected String getConfigurationValue(String attr, String defaultValue) {
+    if (app_configuration.hasPath(attr)) {
+      return app_configuration.getString(attr);
+    }
+    return defaultValue;
+  }
+
+  protected String getEmailAttributeName() {
+    return getConfigurationValue(attributePrefix() + "." + emailAttributeConfigName);
+  }
+
+  protected String getLocaleAttributeName() {
+    return getConfigurationValue(attributePrefix() + "." + localeAttributeConfigName);
+  }
+
+  protected String getFirstNameAttributeName() {
+    return getConfigurationValue(attributePrefix() + "." + firstNameAttributeConfigName);
+  }
+
+  protected String getSecondNameAttributeName() {
+    return getConfigurationValue(attributePrefix() + "." + secondNameAttributeConfigName);
+  }
+
+  protected String getName(OidcProfile oidcProfile) {
+    final String firstNameAttributeName = getFirstNameAttributeName();
+    final String secondNameAttributeName = getSecondNameAttributeName();
+
+    String firstName = "", secondName = "";
+    if (!firstNameAttributeName.isBlank()) {
+      firstName =
+          Objects.toString(oidcProfile.getAttribute(firstNameAttributeName, String.class), "");
+    }
+    if (!secondNameAttributeName.isBlank()) {
+      secondName =
+          Objects.toString(oidcProfile.getAttribute(secondNameAttributeName, String.class), "");
+    }
+    if (!firstName.isBlank() && !secondName.isBlank()) {
+      return String.format("%s %s", firstName, secondName);
+    } else if (!firstName.isBlank()) {
+      return firstName;
+    }
+    return secondName;
+  }
+
+  protected String getLocale(OidcProfile oidcProfile) {
```

Comment:
question (blocking): there's a lot of places where empty strings are passed around and then checked to see if they're blanked. Are the underlying libraries forcing that pattern or could we refactor to use `Optional<String>` instead?